eclipse-theia/theia
# core: fix socket.io endpoint path

Socket.io will try to connect to `domain.com/socket.io` to establish
communication by default. This is problematic when the frontend is
served on an HTTP path that isn't `/`.

Update the path supplied to Socket.io to take the current location into
account.

#### How to test

- Build and run this PR on `localhost:3000`.
- Run the proxy script below like `node index.js 4000 /random-string/ http://localhost:3000/`
- Connect to `localhost:4000/random-string/` with your browser. It should work.

`package.json`:

```json
{
  "name": "http-proxy-test",
  "version": "1.0.0",
  "main": "index.js",
  "license": "MIT",
  "dependencies": {
    "http-proxy": "^1.18.1"
  },
  "devDependencies": {
    "@types/http-proxy": "^1.17.8"
  }
}

```

`index.js`:

```js
const http = require('http')
const httpProxy = require('http-proxy')

let [_node, _script, port, proxyPathPrefix, target] = process.argv

const proxy = httpProxy.createProxyServer({ ws: true, target })
http.createServer((req, res) => {
    if (req.url.startsWith(proxyPathPrefix)) {
        req.url = req.url.substring(proxyPathPrefix.length - 1)
        proxy.web(req, res, undefined, error => console.error(error))
    } else {
        res.statusCode = 404
        res.write(`404: ${req.url}`)
        res.end()
    }
}).on('upgrade', (req, socket, head) => {
    if (req.url.startsWith(proxyPathPrefix)) {
        req.url = req.url.substring(proxyPathPrefix.length - 1)
        proxy.ws(req, socket, head, undefined, error => console.error(error))
    } else {
        socket.end()
    }
}).listen(port)
```

#### Review checklist


#### Reminder for reviewers

- As a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)
